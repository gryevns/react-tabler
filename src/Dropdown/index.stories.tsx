import React from "react";

import Badge from "../Badge";
import Buttons from "../Buttons";
import Card from "../Card";
import Dropdown from "../Dropdown";
import { Calendar, Edit, Settings } from "react-feather";
import Avatar from "../Avatar";

export default {
    title: "Dropdown",
};

export const Button = () => (
    <Buttons.List>
        <Dropdown>
            <Dropdown.Button variant="secondary">Click Me!&nbsp;</Dropdown.Button>
            <Dropdown.Menu>
                <Dropdown.Item>Dropdown link</Dropdown.Item>
                <Dropdown.Item active>Dropdown link</Dropdown.Item>
                <Dropdown.Divider />
                <Dropdown.Item disabled>Dropdown link</Dropdown.Item>
            </Dropdown.Menu>
        </Dropdown>
        <Dropdown>
            <Dropdown.Button variant="secondary">
                <Calendar size={18} />
                &nbsp;
            </Dropdown.Button>
            <Dropdown.Menu>
                <Dropdown.Header>Header</Dropdown.Header>
                <Dropdown.Item>
                    <Settings size={15} className="mr-2" />
                    Dropdown link
                </Dropdown.Item>
                <Dropdown.Item>
                    <Edit size={15} className="mr-2" />
                    Dropdown link
                    <Badge variant="primary" className="ml-auto">
                        12
                    </Badge>
                </Dropdown.Item>
            </Dropdown.Menu>
        </Dropdown>
    </Buttons.List>
);

export const Dark = () => (
    <Dropdown>
        <Dropdown.Button variant="secondary">Dark Dropdown</Dropdown.Button>
        <Dropdown.Menu dark>
            <Dropdown.Header>Header</Dropdown.Header>
            <Dropdown.Item>Dropdown link</Dropdown.Item>
            <Dropdown.Item>Dropdown link</Dropdown.Item>
        </Dropdown.Menu>
    </Dropdown>
);

export const CardContent = () => (
    <Dropdown>
        <Dropdown.Button variant="secondary">Card Content</Dropdown.Button>
        <Dropdown.Menu card>
            <Card className="d-flex flex-column">
                <Card.Image
                    top
                    src="https://picsum.photos/id/562/600/200"
                    alt="How do you know she is a witch?"
                />
                <Card.Body>
                    <Card.Title>How do you know she is a witch?</Card.Title>
                    <div className="text-muted">
                        Are you suggesting that coconuts migrate? No, no, no! Yes,
                        yes. A bit. But she's got a wart. You ...
                    </div>
                    <div className="d-flex align-items-center pt-5 mt-auto">
                        <Avatar imageURL="https://picsum.photos/id/128/128" />
                        <div className="ml-3">
                            <a href="#" className="text-body">
                                Maryjo Lebarree
                            </a>
                            <small className="d-block text-muted">3 days ago</small>
                        </div>
                    </div>
                </Card.Body>
            </Card>
        </Dropdown.Menu>
    </Dropdown>
);
