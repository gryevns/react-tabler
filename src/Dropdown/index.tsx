import React, { useEffect, useRef } from "react";
import cn from "classnames";
import Button from "../Button";
import Nav from "../Nav";
import { Variant } from "../common";

export const DropdownContext = React.createContext({
    dropdownVisible: false,
    setDropdownVisible: (visible: boolean): any => {},
});

const DropdownHeader = ({ children }) => (
    <span className="dropdown-header">{children}</span>
);

export interface DropdownItemProps {
    className?: string;
    children: React.ReactNode;
    active?: boolean;
    disabled?: boolean;
}

const DropdownItem = ({
    className,
    children,
    active,
    disabled,
}: DropdownItemProps) => {
    const classes = cn("dropdown-item", { active, disabled }, className);
    return (
        <a
            className={classes}
            href="#"
            onClick={(e: React.MouseEvent) => e.preventDefault()}
        >
            {children}
        </a>
    );
};

export interface DropdownMenuProps {
    children: React.ReactNode;
    className?: string;
    arrow?: boolean;
    card?: boolean;
    position?: "right" | "left";
    style?: React.CSSProperties;
    dark?: boolean;
}

const DropdownMenu = ({
    children,
    className,
    arrow,
    card,
    position,
    style,
    dark,
}: DropdownMenuProps) => {
    const { dropdownVisible } = React.useContext(DropdownContext);
    const optionalClasses = {
        show: dropdownVisible,
        "dropdown-menu-arrow": arrow,
        "dropdown-menu-card": card,
        "dropdown-menu-right": position === "right",
        "dropdown-menu-left": position === "left",
        "bg-dark text-white": dark,
    };
    const classes = cn("dropdown-menu", optionalClasses, className);
    return (
        <div className={classes} style={style}>
            {children}
        </div>
    );
};

export interface DropdownProps {
    className?: string;
    children: React.ReactNode;
    unfocus?: boolean;
}

const Dropdown = ({ className, children, unfocus = true }: DropdownProps) => {
    const ref = useRef(null);
    const [dropdownVisible, setDropdownVisible] = React.useState(false);
    const classes = cn("dropdown", className);

    function handleClickOutside(event) {
        if (ref.current && !ref.current.contains(event.target) && dropdownVisible) {
            setDropdownVisible(false);
        }
    }

    if (unfocus) {
        // better name?
        useEffect(() => {
            document.addEventListener("mousedown", handleClickOutside);
            return () => {
                document.removeEventListener("mousedown", handleClickOutside);
            };
        });
    }

    return (
        <DropdownContext.Provider value={{ dropdownVisible, setDropdownVisible }}>
            <div ref={ref} className={classes}>
                {children}
            </div>
        </DropdownContext.Provider>
    );
};

export interface DropdownButtonProps {
    children?: React.ReactNode;
    className?: string;
    variant?: Variant;
}

const DropdownButton = ({ children, className, ...props }: DropdownButtonProps) => {
    const { dropdownVisible, setDropdownVisible } = React.useContext(
        DropdownContext,
    );
    const onClick = () => setDropdownVisible(!dropdownVisible);
    const classes = cn("dropdown-toggle", className);
    return (
        <Button
            onClick={onClick}
            className={classes}
            {...props}
            data-toggle="dropdown"
            aria-expanded={dropdownVisible}
        >
            {children}
        </Button>
    );
};

export interface DropdownLinkProps {
    children: React.ReactNode;
    className?: string;
}

const DropdownLink = ({ children, className, ...props }: DropdownLinkProps) => {
    const { dropdownVisible, setDropdownVisible } = React.useContext(
        DropdownContext,
    );
    const onClick = (e: React.MouseEvent) => {
        setDropdownVisible(!dropdownVisible);
        e.preventDefault();
    };
    const classes = cn(className);
    return (
        <Nav.Link
            onClick={onClick}
            className={classes}
            {...props}
            data-toggle="dropdown"
            aria-expanded={dropdownVisible}
        >
            {children}
        </Nav.Link>
    );
};

const DropdownDivider = () => <div className="dropdown-divider" />;

Dropdown.Button = DropdownButton;
Dropdown.Link = DropdownLink;
Dropdown.Menu = DropdownMenu;
Dropdown.Item = DropdownItem;
Dropdown.Divider = DropdownDivider;
Dropdown.Header = DropdownHeader;

export default Dropdown;
