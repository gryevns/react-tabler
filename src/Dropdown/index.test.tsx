import React from "react";
import { shallow, mount } from "enzyme";

import Dropdown from "./index";

describe("Dropdown", () => {
    it("should match snapshot", () => {
        const component = shallow(
            <Dropdown>
                <Dropdown.Button variant="secondary">Click Me!</Dropdown.Button>
                <Dropdown.Menu>
                    <Dropdown.Item>Dropdown link</Dropdown.Item>
                    <Dropdown.Item>Dropdown link</Dropdown.Item>
                </Dropdown.Menu>
            </Dropdown>,
        );
        expect(component).toMatchSnapshot();
    });
    it("should match link snapshot", () => {
        const component = shallow(
            <Dropdown>
                <Dropdown.Link>Click Me!&nbsp;</Dropdown.Link>
                <Dropdown.Menu arrow>
                    <Dropdown.Item>Dropdown link</Dropdown.Item>
                    <Dropdown.Item>Dropdown link</Dropdown.Item>
                </Dropdown.Menu>
            </Dropdown>,
        );
        expect(component).toMatchSnapshot();
    });
    it("should render dropdown on click", () => {
        const component = mount(
            <Dropdown>
                <Dropdown.Button variant="secondary">Click Me!</Dropdown.Button>
                <Dropdown.Menu>
                    <Dropdown.Item>Dropdown link</Dropdown.Item>
                    <Dropdown.Item>Dropdown link</Dropdown.Item>
                </Dropdown.Menu>
            </Dropdown>,
        );
        expect(component.find(".dropdown-menu").prop("className")).not.toContain(
            "show",
        );
        component.find("button").simulate("click");
        expect(component.find(".dropdown-menu").prop("className")).toContain("show");
    });
});
