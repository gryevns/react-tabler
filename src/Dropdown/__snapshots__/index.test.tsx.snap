// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should match snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "dropdownVisible": false,
      "setDropdownVisible": [Function],
    }
  }
>
  <div
    className="dropdown"
  >
    <DropdownButton
      variant="secondary"
    >
      Click Me!
    </DropdownButton>
    <DropdownMenu>
      <DropdownItem>
        Dropdown link
      </DropdownItem>
      <DropdownItem>
        Dropdown link
      </DropdownItem>
    </DropdownMenu>
  </div>
</ContextProvider>
`;
